src/libclipper/src/selection_policies.cpp

```diff
@@ -1,344 +1,407 @@
-
-#include <cmath>
-#include <memory>
+#include <float.h>
+#include <math.h>
+#include <functional>
+#include <iostream>
+#include <random>
 #include <stdexcept>
 #include <string>
 #include <utility>
 #include <vector>
 
+#include <boost/archive/binary_iarchive.hpp>
+#include <boost/archive/binary_oarchive.hpp>
+#include <boost/serialization/string.hpp>
+#include <boost/serialization/unordered_map.hpp>
+#include <boost/serialization/utility.hpp>
+
 #include <clipper/datatypes.hpp>
-#include <clipper/selection_policy.hpp>
+#include <clipper/selection_policies.hpp>
 #include <clipper/util.hpp>
 
 namespace clipper {
 
-VersionedModelId NewestModelSelectionPolicy::initialize(
-    const std::vector<VersionedModelId>& candidate_models) {
-  // TODO: IMPLEMENT
-  assert(candidate_models.size() > 0);
-  return candidate_models.front();
+// ********
+// * EXP3 *
+// ********
+// Descripton: Single Model Selection Policy, iteratively update weights
+Exp3State Exp3Policy::initialize(
+    const std::vector<VersionedModelId>& candidate_models_) {
+  // Exp3: sum of weights; each model has "weight", "max loss"
+  Map map(candidate_models_.size(), &versioned_model_hash);
+  for (VersionedModelId id : candidate_models_) {
+    ModelInfo info = {{"weight", 1.0}, {"max loss", 0.0}};
+    map.insert({id, info});
+  }
+  return make_pair(map.size() * 1.0, map);
 }
 
-VersionedModelId NewestModelSelectionPolicy::add_models(
-    VersionedModelId state, std::vector<VersionedModelId> new_models) {
-  UNUSED(state);
-  return new_models.front();
+Exp3State Exp3Policy::add_models(
+    Exp3State state, const std::vector<VersionedModelId>& new_models) {
+  // Give new models average weight from old models
+  auto avg = state.first / state.second.size();
+  for (VersionedModelId id : new_models) {
+    ModelInfo info = {{"weight", avg}, {"max loss", 0.0}};
+    state.second.insert({id, info});
+    state.first += avg;
+  }
+  return state;
 }
 
-long NewestModelSelectionPolicy::hash_models(
-    const std::vector<VersionedModelId>& candidate_models) {
-  UNUSED(candidate_models);
-  return 0;
+VersionedModelId Exp3Policy::select(Exp3State state,
+                                    std::vector<VersionedModelId>& models) {
+  // Helper function for randomly drawing an arm based on its normalized weight
+  auto rand_num = rand() % 1;
+  VersionedModelId selected_model;
+  auto it = models.begin();
```

Comment:
What about the case where `models` is empty? You should log an error message and select prediction tasks should return an empty vector.